[core] RestCatalog: set up paged list tables/views/partitions

XiaoHongbo-Hope opened the pull request:
<!-- Please specify the module before the PR name: [core] ... or [flink] ... -->

[core] RestCatalog: set up paged list tables/views/partitions

<!-- Linking this pull request to the issue -->


<!-- What is the purpose of the change -->
Check Iceberg & Unity Catalog,  they support paged list for tables & views both, so suggest adding paged list method into paimon 

### Tests

<!-- List UT and IT cases to verify this change -->

### API and Format

<!-- Does this change affect API or storage format -->

### Documentation

<!-- Does this change introduce a new feature -->

sundapeng commented on paimon-open-api/src/main/java/org/apache/paimon/open/api/RESTCatalogController.java:
```diff
@@ -226,8 +228,47 @@ public AlterDatabaseResponse alterDatabase(
-        return new ListTablesResponse(ImmutableList.of("user"));
+            @PathVariable String prefix,
+            @PathVariable String database,
+            @PathVariable Integer maxResults,
```
Please update API spec `rest-catalog-open-api.yaml`

XiaoHongbo-Hope replied:
ok, thanks

sundapeng commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:
```diff
@@ -560,16 +690,34 @@ public void markDonePartitions(Identifier identifier, List<Map<String, String>>
+                            identifier.getTableName(),
+                            queryParams);
+                }
+            } while (pageToken != null);
```
It is better to check `pageToken.isNotEmpty()`

sundapeng commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:
```diff
@@ -257,16 +270,133 @@ public void alterDatabase(String name, List<PropertyChange> changes, boolean ign
+                                                    Identifier.create(
+                                                            databaseName,
+                                                            getTableResponse.getName());
+                                            try {
```
Is it possible to reuse some utility methods, such as `getTable(Identifier identifier)`

sundapeng commented:
Looks good overall, but minor comments.

JingsongLi commented on paimon-common/src/main/java/org/apache/paimon/PagedList.java:
```diff
@@ -0,0 +1,42 @@
+
+/** Paged List which supports request data from page streaming. */
+public class PagedList<T> implements Serializable {
+    private final List<T> pagedLists;
```
Please add ser id for Serializable class.

XiaoHongbo-Hope replied:
Since we will hold some un Serializable data in the PagedList, such as Table/View,  I remove the Serializable implements.

JingsongLi commented on paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java:
```diff
@@ -130,6 +131,44 @@ void alterDatabase(String name, List<PropertyChange> changes, boolean ignoreIfNo
+     *     token
+     * @throws DatabaseNotExistException if the database does not exist
+     */
+    PagedList<String> listTablesPaged(String databaseName, Integer maxResults, String pageToken)
```
`@Nullable` for nullable arguements.

XiaoHongbo-Hope replied:
ok, thanks

JingsongLi reviewed (commented):
Thanks @XiaoHongbo-Hope for the contribution! Pretty good!

XiaoHongbo-Hope commented:
Can you please help review again? I have fixed all the comment，optimized the test case assertion and cleaned code for the mock server. Note the catalog client logic should be same as last review.
@sundapeng @JingsongLi @jerry-024

jerry-024 commented on paimon-common/src/main/java/org/apache/paimon/PagedList.java:
```diff
@@ -0,0 +1,41 @@
+
+/** Paged List which supports request data from page streaming. */
+public class PagedList<T> {
+    private final List<T> pagedLists;
```
just pagedList or data?

XiaoHongbo-Hope replied:
will update pagedLists to pagedList.

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:
```diff
@@ -240,8 +254,69 @@ public List<String> listTables(String databaseName) throws DatabaseNotExistExcep
+    protected PagedList<Table> listTableDetailsPagedImpl(
+            String databaseName, Integer maxResults, String pageToken)
+            throws DatabaseNotExistException {
+        PagedList<String> pagedTableNames = listTablesPaged(databaseName, maxResults, pageToken);
```
Is it possible to get table detail directly?

XiaoHongbo-Hope replied:
will get table details directly in RestCatalog, this is the default implements for other catalog.

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java:
```diff
@@ -130,6 +133,45 @@ void alterDatabase(String name, List<PropertyChange> changes, boolean ignoreIfNo
+     *     the result. If maxResults is not specified or set to 0, will return the default number of
+     *     max results.
+     * @param pageToken Optional parameter indicating the next page token allows list to be start
+     *     from a specific point.
```
from a specific table name?

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java:
```diff
@@ -326,6 +385,46 @@ default List<String> listViews(String databaseName) throws DatabaseNotExistExcep
+    default PagedList<String> listViewsPaged(
+            String databaseName, @Nullable Integer maxResults, @Nullable String pageToken)
+            throws DatabaseNotExistException {
+        return new PagedList<>(listViews(databaseName), null);
```
Why is there a default implementation？

XiaoHongbo-Hope replied:
just follow the behavior of listViews in interface Catalog, maybe because most catalog does not support view functions.

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/catalog/Catalog.java:
```diff
@@ -326,6 +385,46 @@ default List<String> listViews(String databaseName) throws DatabaseNotExistExcep
+    default PagedList<View> listViewDetailsPaged(
+            String databaseName, @Nullable Integer maxResults, @Nullable String pageToken)
+            throws DatabaseNotExistException {
+        return new PagedList<>(Collections.emptyList(), null);
```
Why is there a default implementation？

sundapeng commented:
LGTM, thank @XiaoHongbo-Hope

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/rest/HttpClient.java:
```diff
@@ -191,7 +209,23 @@ private static RequestBody buildRequestBody(String body) throws JsonProcessingEx
+        return getRequestUrl(path, null);
+    }
+
+    private String getRequestUrl(String path, Map<String, String> queryParams) {
```
Is there any test case for this method？

XiaoHongbo-Hope replied:
added, and test case of catalog will cover this too

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:
```diff
@@ -257,16 +271,117 @@ public void alterDatabase(String name, List<PropertyChange> changes, boolean ign
+                                queryParams,
+                                ListTablesResponse.class,
+                                restAuthFunction);
+                if (Objects.nonNull(response)) {
```
Could be response null here?

XiaoHongbo-Hope replied:
I think we can gain more and loss less by double checking

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:
```diff
@@ -257,16 +271,117 @@ public void alterDatabase(String name, List<PropertyChange> changes, boolean ign
                             restAuthFunction);
-            if (response.getTables() != null) {
-                return response.getTables();
+            if (Objects.nonNull(response) && Objects.nonNull(response.getTables())) {
```
Could be response null here?

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:
```diff
@@ -257,16 +271,117 @@ public void alterDatabase(String name, List<PropertyChange> changes, boolean ign
+                            queryParams,
+                            ListTableDetailsResponse.class,
+                            restAuthFunction);
+            if (Objects.nonNull(response) && Objects.nonNull(response.getTableDetails())) {
```
Could be response null here?

jerry-024 commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:
```diff
@@ -560,16 +675,34 @@ public void markDonePartitions(Identifier identifier, List<Map<String, String>>
+                                queryParams,
+                                ListPartitionsResponse.class,
+                                restAuthFunction);
+                if (Objects.nonNull(response)) {
```
Could be response null here?

XiaoHongbo-Hope commented on paimon-common/src/main/java/org/apache/paimon/PagedList.java:
```diff
@@ -0,0 +1,41 @@
+
+/** Paged List which supports request data from page streaming. */
+public class PagedList<T> {
+    private final List<T> pagedList;
```
👌

XiaoHongbo-Hope commented on paimon-common/src/main/java/org/apache/paimon/PagedList.java:
```diff
@@ -0,0 +1,41 @@
+        return this.pagedList;
+    }
+
+    public String getNextPageToken() {
```
👌

XiaoHongbo-Hope commented on paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalogOptions.java:
```diff
@@ -85,4 +85,10 @@ public class RESTCatalogOptions {
                     .withDescription("Whether to support data token provided by the REST server.");
+
+    public static final ConfigOption<Integer> REST_PAGE_MAX_RESULTS =
+            ConfigOptions.key("rest.page.max.results")
```
👌

JingsongLi reviewed (approved):
Looks good to me!
